lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -158,6 +159,99 @@ static SmallVector<int64_t> computeDimsOrderForMoveDim(int64_t srcDimInt,
   return dimsOrder;
 }
 
+static bool
+rewriteEquationWithEllipsisSlicing(std::string &equation,
+                                   SmallVector<int64_t> &inputRanks) {
+  // split equation into input and result
+  size_t arrowPos = equation.find("->");
+  if (arrowPos == std::string::npos) {
+    return false;
+  }
+  std::string inputStr = equation.substr(0, arrowPos);
+  std::string resultStr = equation.substr(arrowPos + 2);
+
+  // split input into tokens
+  llvm::SmallVector<std::string> inputTokens;
+  size_t start = 0;
+  size_t end = 0;
+  std::set<char> usedTokens;
+  while (end < inputStr.size()) {
+    end = inputStr.find(",", start);
+    if (end == std::string::npos) {
+      end = inputStr.size();
+    }
+    std::string token = inputStr.substr(start, end - start);
+    inputTokens.push_back(token);
+    start = end + 1;
+  }
+  if (inputTokens.size() != inputRanks.size()) {
+    return false;
+  }
+
+  // find the rank which ellipsis represents, and max ellipsis rank because a
+  // tensor can be broadcasted
+  SmallVector<int64_t> ellipsisRanks;
+  int maxEllipsisRank = 0;
+  for (const auto &[token, inputRank] : llvm::zip(inputTokens, inputRanks)) {
+    int explictRank = 0;
+    for (auto c : token) {
+      if (std::isalpha(c)) {
+        usedTokens.insert(c);
+        explictRank++;
+      } else if (c == '.' || c == ' ') {
+        continue;
+      } else {
+        return false;
+      }
+    }
+    int ellipsisRank = inputRank - explictRank;
+    if (ellipsisRank > maxEllipsisRank) {
+      maxEllipsisRank = ellipsisRank;
+    }
+    if (ellipsisRank < 0) {
+      return false;
+    }
+    ellipsisRanks.push_back(inputRank - explictRank);
+  }
+  auto ellipsisToken = [&usedTokens, &maxEllipsisRank]() {
+    std::string token;
+    int num = maxEllipsisRank;
+    char lastChar = 'a';
+    while (num--) {
+      while (usedTokens.find(lastChar) != usedTokens.end()) {
+        lastChar++;
+      }
+      token.push_back(lastChar);
+      lastChar++;
+    }
+    return token;
```

Comment:
nit: this does not need to be a lambda expression. I think `token` can just be replaced with `ellipsisToken` and the lambda expression can be removed.

Reply:
I have already made the modifications according to your suggestions.